### ISSUE TITLE
Synchronising session/bookmarks/history

### ISSUE THREAD

**Issue by bijang** [atlas-engineer/nyxt#631]
Currently I mainly use Firefox with a few addons including
https://tab-session-manager.sienori.com/
https://github.com/piroor/treestyletab
https://www.xbrowsersync.org/
It's an incomplete solution, not everything is synced yet, but I chose this because it was the best combo for my demands at this time.
The  next browser can do everything already with relation to advanced bookmark management, history as tree etc

it does raise a few questions for me, (forgive me I don't have next installed on this machine at present). 

- How is history managed in next, does it erase after a time or can it be erased? 
- This is where history and 'session' overlap, ie history is made up of sessions. Can one 'save' a session (meaning the open windows/tabs), to then come back to the same, or even a previous session (in the way managed by the tab session manager extension)? 
In this extension, the session is tagged automatically with metadate including time and date, number  of windows open and number of tabs within windows open, as well as optionally giving other tags or titles to the session. [I understand the concept of tabs in next is different]. 
Please have a look at tab session extension if you have time, it can be easier than my descibing it. 
Or am I barking up the wrong tree, and the way the tab paradigm works in next and the history, with ability to fuzzy search between places, provides a 'fluidity' in a way that means there is no such thing as a well defined 'session' you may have in a traditional browser (again I havn't used next enough yet myself to answer these questions). 
I think all this probably depends on how you define a session. I'm glad there is tree history in next at least allow some intropection into the work flow that went before. Is the history tree populated as a plain text list or do they conisit of working hyperlinks? (if not would this be possible?)
- Synchronisation, of bookmarks and related metadata, history and session (whatever this may be defined as); ie portability. So if I move to a different laptop can I access the history and bookmarks etc, and therefore modify them, and have these synced up. Could this be done without the need for stting up a central server instance (like xbrowsersync)? Would the data set up be like a git where can push and pull into it? The way pass does for password management. Would security factor into it (in terms of holding the browser hisotry and bookmarks in an encrypted file)

I think this last one is important as next gets more popular and esp when next gets it's inevitable companion mobile app ;)

Apologies is so long, I hope this is ok as an issue, or perhaps you want me to put it into a few seperate issues?

**Comment by tviti** [comment]
For your last bullet: I wound up making the bookmark dir a git repo, and
rolled a few commands for pushing/pulling. The implementation is pretty
shitty, and every so often I have to go in and manually merge (when I
forget to sync for a while), but maybe this will give you some ideas.

Note that the session file is also stored in the bookmark dir, so you can
check it in if you want session persistence too (I don't bother tho).

Another thought I've had, which I haven't tried, is to add the bookmark dir
as a directory in Syncthing.

Tbh, I'm not sure if I think that syncing should be Next' responsibility,
since there are so many ways to do it.

**Comment by Ambrevar** [comment]
> Currently I mainly use Firefox with a few addons including
> https://tab-session-manager.sienori.com/
> https://github.com/piroor/treestyletab
> https://www.xbrowsersync.org/
> It's an incomplete solution, not everything is synced yet, but I chose this because it was the best combo for my demands at this time.
> The  next browser can do everything already with relation to advanced bookmark management, history as tree etc
>
> it does raise a few questions for me, (forgive me I don't have next installed on this machine at present). 
>
> - How is history managed in next, does it erase after a time or can it be erased? 

It's never erased by default.
There is the "delete-history-entry" command which accepts multiple
selections.
So if you want to clean up the full history, run

  M-x delete-history-entry return M-a return

Or simply delete the file.

> - This is where history and 'session' overlap, ie history is made up
> of sessions. Can one 'save' a session (meaning the open windows/tabs),
> to then come back to the same, or even a previous session (in the way
> managed by the tab session manager extension)?

> In this extension,

Which extension?

> the session is tagged automatically with metadate including time and
> date, number of windows open and number of tabs within windows open,
> as well as optionally giving other tags or titles to the session. [I
> understand the concept of tabs in next is different].

Note that sessions store all buffers.

I've recently updated session support to make it easier to support
multiple sessions.
For instance you could start next like this

  next --session session-$(date '+%F-%T')

and every start of Next will create a new session name after the date it
was started.

If you want to recycle the previous session into a new file, I suppose
that a tiny function in your init.lisp could do.  Contributions are welcome!

> Please have a look at tab session extension if you have time, it can
> be easier than my descibing it.

Which one?  Tab Session Manager?

> Or am I barking up the wrong tree, and the way the
> tab paradigm works in next and the history, with ability to fuzzy
> search between places, provides a 'fluidity' in a way that means there
> is no such thing as a well defined 'session' you may have in a
> traditional browser (again I havn't used next enough yet myself to
> answer these questions).  I think all this probably depends on how you
> define a session.

For now a Next session is simply the list of all open buffers (with
their histories).  We are not saving the windows, because I'm not sure
that would be too useful at the moment.

However when we implement #503, this might be useful information indeed.

> I'm glad there is tree history in next at least
> allow some intropection into the work flow that went before. Is the
> history tree populated as a plain text list or do they conisit of
> working hyperlinks? (if not would this be possible?)

Sorry, I didn't understand your question :p

The history tree is a data structure that holds a tree of all URLs that
were visited in the current buffer.

This history tree is persisted in the plain-text session file.

The `buffer-history-tree` command gives you a visualization of this tree
with clickable links.

The `history-all-query` allows you to fuzzy-search among all the links
and to load the selected link in the current buffer.


> - Synchronisation, of bookmarks and related metadata, history and
> session (whatever this may be defined as); ie portability. So if I
> move to a different laptop can I access the history and bookmarks etc,
> and therefore modify them, and have these synced up. Could this be
> done without the need for stting up a central server instance (like
> xbrowsersync)? Would the data set up be like a git where can push and
> pull into it? The way pass does for password management. Would
> security factor into it (in terms of holding the browser hisotry and
> bookmarks in an encrypted file)

To add to what Tviti said, I think the easiest way is to use Git for synchronization.

In the near future, I'd like to add transparent GPG supports so that
bookmarks etc. get encrypted on the fly.

We could also add Git synchronization support following up on Tviti's
implementation.
This is a bit tricky to get right if we need to handle conflicts.  The
ideal solution here would be a Magit-like interface in Next ;)

In the longer term, Atlas Engineer could provide Git repositories so
that bookmark syncing works out of the box.

> Apologies is so long, I hope this is ok as an issue, or perhaps you want me to put it into a few seperate issues?

No problem for now, we can open separate issues when the time comes, or
when anyone decides to work on them! :)

**Comment by bijang** [comment]
Many thanks for your reply. Apologies for the lack of clarity above I was referring predominantly to the https://tab-session-manager.sienori.com/ when I mentioned a browser extension. On the fly encrypting and git sync sounds like great idea.

**Comment by bijang** [comment]
libp2p?.. 
this is clearly a post 2.x release issue but I have one possibly crazy idea I need to float here (before I forget) [bear in mind I'm (embarrasingly) not a coder and I don't know if this is feasable or too much work] in relation to syncing (history, bookmarks, password store etc) data (as discussed above) could, in the vein of matrix, it be possible to have the option to sync p2p OR via server (self hosted or otherwise)... p2p utilising a libp2p implementation? ....just a thought, it would be supercool if that was possible, taking away the hassle (and security problems) of a server and lowering the threshold to adoption while also allowing the option for server based storage for eg companies, groups or individuals with certain demands...
again I release this is a loooong way down the priority list at the moment, but I had to let it air...
can I now thank you for the incredible amount of work you're putting into this fantastic project and ask will you be doing another kickstarter type campaign (maybe post 2.0) or have a way to donate funds [if this were acceptable to your business?] (mainly because I have no coding ability and all I possibly could donate is a small amount of monies)

**Comment by Ambrevar** [comment]
libp2p?  Why not! :)
We would probably need a dedicated library, not sure there is one for
Common Lisp already.

**Comment by Ambrevar** [comment]
Since commit e5cedaf9de1181825e9ec1362af1a3fd42586f44 Next has transparent support for GPG encrypted files.
This means that if you tell Next to store your bookmarks as "bookmarks.lisp.gpg" instead of "bookmarks.lisp", it will encrypt them transparently.

**Comment by jmercouris** [comment]
Yay! Now maybe I can put my bookmarks in a public repository! :-)

**Comment by bijang** [comment]
well done! and thank you for this! it's a (dare I say) unique, unbelievably important and foundational development that people can build on!

**Comment by bijang** [comment]
is history and session involved in the encyption the way it's said to be on the changelog for 2.0? :)
and is git synch coming?

**Comment by Ambrevar** [comment]
You can persist your history and session as a GPG-encrypted file.

Git sync might come some day, nothing has been done in that direction yet!

**Comment by Ambrevar** [comment]
@ykgoon Not sure what you mean, bookmarks are already stored locally and yes, you can sync them with any service you like.

That said, for the sake of your privacy, you should not synchronize unencrypted bookmarks with an untrusted third-party (this includes Dropbox).

**Comment by ykgoon** [comment]
> That said, for the sake of your privacy, you should not synchronize unencrypted bookmarks with an untrusted third-party (this includes Dropbox).

Agreed. So as long as its encrypted it'd be fine with me without a centralized sync service.
